snuba/snuba_migrations/spans/0001_spans_v1.py

```diff
@@ -0,0 +1,99 @@
+local_table_name = "spans_local"
+dist_table_name = "spans_dist"
+
+UNKNOWN_SPAN_STATUS = 2
+
+columns: List[Column[Modifiers]] = [
+    Column("project_id", UInt(64)),
+    Column("transaction_id", UUID(Modifiers(nullable=True))),
+    Column("trace_id", UUID()),
+    Column("span_id", UInt(64)),
+    Column("parent_span_id", UInt(64, Modifiers(nullable=True))),
+    Column("segment_id", UInt(64)),
+    Column("is_segment", UInt(8)),
+    Column("segment_name", String(Modifiers(default="''"))),
+    Column("start_timestamp", DateTime()),
+    Column("end_timestamp", DateTime()),
+    Column("timestamp", DateTime(Modifiers(materialized="end_timestamp"))),
+    Column("duration", UInt(32)),
+    Column("exclusive_time", Float(64)),
+    Column("op", String(Modifiers(low_cardinality=True))),
+    Column("group", UInt(64)),
+    Column("span_status", UInt(8)),
+    Column("span_kind", String(Modifiers(low_cardinality=True))),
+    Column("description", String()),
+    Column("status", UInt(32, Modifiers(nullable=True))),
+    Column("module", String(Modifiers(low_cardinality=True, nullable=True))),
+    Column("action", String(Modifiers(low_cardinality=True, nullable=True))),
+    Column("domain", String(Modifiers(nullable=True))),
+    Column("platform", String(Modifiers(low_cardinality=True, nullable=True))),
+    Column("user", String(Modifiers(nullable=True))),
+    Column("tags", Nested([("key", String()), ("value", String())])),
+    Column(
+        "measurements",
+        Nested(
+            [("key", String(Modifiers(low_cardinality=True))), ("value", Float(64))]
+        ),
+    ),
+    Column("partition", UInt(16)),
+    Column("offset", UInt(64)),
+    Column("retention_days", UInt(16)),
+    Column("deleted", UInt(8)),
+]
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    blocking = False
+
+    def forwards_ops(self) -> Sequence[SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=storage_set_name,
+                table_name=local_table_name,
+                columns=columns,
+                engine=table_engines.ReplacingMergeTree(
+                    order_by="(project_id, segment_name, segment_id, timestamp)",
+                    version_column="deleted",
+                    partition_by="(retention_days, toMonday(timestamp))",
+                    sample_by="segment_id",
+                    settings={"index_granularity": "8192"},
+                    storage_set=storage_set_name,
+                    ttl="timestamp + toIntervalDay(retention_days)",
+                ),
+                target=OperationTarget.LOCAL,
+            ),
+            operations.CreateTable(
+                storage_set=storage_set_name,
+                table_name=dist_table_name,
+                columns=columns,
+                engine=table_engines.Distributed(
+                    local_table_name=local_table_name,
+                    sharding_key="cityHash64(segment_id)",
```

Comment:
> From https://github.com/getsentry/snuba/pull/4081#discussion_r1178528521, Fillipo: Why not sharding by trace id so that we can run complex subqueries on the same trace locally ?

That is a good point. If we need to support querying by `trace_id`, sharding by `trace_id` would be better. Also, by virtue of a `segment_id` belonging to a `trace_id`, a given `segment_id` would always be on a given shard.I will modify it.